### ISSUE TITLE
Implementation of SSDMultiBoxLoss

### ISSUE THREAD

**Issue by caiqi** [dmlc/gluon-cv#226]
The classification loss is divided by the number of positive samples. Would it be better to divide by the total number of positive and negative samples?

https://github.com/dmlc/gluon-cv/blob/master/gluoncv/loss.py#L150

And it does not take -1 in cls_target into consideration, where -1 often means ignoring the loss.

**Comment by zhreshold** [comment]
Divide by positive sample == divide by (pos+neg) samples / 4, which is controlled by OHEM sampler.
You can choose whatever way to normalize the loss, with another weight to control it further. 

For -1 samples, they are ignored in loss, and same here for the normalization factor.

**Comment by zhreshold** [comment]
nd.pick will pick according to ct anyway, 0 is the first elem, -1 is the last elem. So it won't ignore the samples, loss are actually masked out by element-wise weight during loss calculation.

**Comment by caiqi** [comment]
I'm sorry I didn't make myself clear. 

> nd.pick will pick according to ct anyway, 0 is the first elem, -1 is the last elem. 

I found that -1 is also the first elem.  This is the reproducing code.

```
from mxnet import nd
import mxnet as mx
print(mx.__version__)
mx.random.seed(12)
data = nd.random.uniform(shape=(1, 5, 10))
ct = (nd.random.uniform(shape=(1, 5)) * 10).astype("int8")
ct[0, 0] = -1
selected_loss = nd.pick(data, index=ct, axis=-1)
print(data)
print(ct)
print(selected_loss)
```

This is the output:

```
1.2.1

[[[0.9417475  0.5620318  0.3910588  0.7953725  0.05816786 0.91905856
   0.02507077 0.22761478 0.53458214 0.9575423 ]
  [0.4423109  0.21397245 0.6902101  0.4342196  0.3084805  0.8455637
   0.3634608  0.36192498 0.64434606 0.6033084 ]
  [0.29353532 0.4326603  0.7384209  0.7709029  0.01085234 0.23640934
   0.5513674  0.9169604  0.22354734 0.08366717]
  [0.540642   0.33259252 0.08773082 0.19753866 0.25946602 0.90321404
   0.41612142 0.7774379  0.5059712  0.7243525 ]
  [0.13981043 0.23955366 0.76991767 0.19341262 0.4641797  0.08688452
   0.8271064  0.8246709  0.5130552  0.13239743]]]
<NDArray 1x5x10 @cpu(0)>

[[-1  2  1  1  0]]
<NDArray 1x5 @cpu(0)>

[[0.9417475  0.6902101  0.4326603  0.33259252 0.13981043]]
<NDArray 1x5 @cpu(0)>
```
The first item in ct is -1, nd.pick select 0.9417475 rather than 0.9575423. In most detection model, 0 means background class. So I think nd.pick takes -1 as 0 and thus treats _ignore_ in _ct_ as _background_.

I originally thought change this line:

https://github.com/dmlc/gluon-cv/blob/master/gluoncv/loss.py#L146

from 
```
rank = (cls_loss * (pos - 1)).argsort(axis=1).argsort(axis=1)
```
to 
```
rank = (cls_loss * ((ct !=0) - 1)).argsort(axis=1).argsort(axis=1)
```
would be better. Since it will mask both positive(>0) and ignored(-1) loss as zeros when finding hard negative samples.

**Comment by zhreshold** [comment]
okay, I can confirm that pick does not support negative index, it just clip the index to (0, len) not wrapping them. 
However, for the loss in this case, we will always use a mask, where the mask is `ct >=0` for example, so the final loss is `selected_loss * (ct >= 0)`, effectively the -1 samples are masked out.

**Comment by caiqi** [comment]
Thanks for the explanation!
